### ISSUE TITLE
Automated builds/releases

### ISSUE THREAD

**Issue by JoeBiellik** [TV-Rename/tvrename#187]
@MarkSummerville mentioned in #185 the need for an easier build/deploy process. From what I can see, currently, NSIS is manually used to build an installer package from a compiled release build.

Version numbers across the project seem sporadic, [Version.cs](https://github.com/TV-Rename/tvrename/blob/master/TVRename%23/App/Version.cs) seems outdated, [AssemblyInfo.cs](https://github.com/TV-Rename/tvrename/blob/master/TVRename%23/Properties/AssemblyInfo.cs)'s ``AssemblyVersion`` is wrong and ``AssemblyFileVersion`` is used to determine the config directory (see #186).
Finally, the NSIS scripts are a little outdated and incorrectly report that .NET Framework 2.0 is required when the project is built against 4.0 and include dead download links.

I've had great success using a free [AppVeyor](https://www.appveyor.com/) account to automatically pull, build and push to a GitHub release, so that's what I would strongly suggest.

I'll happily make a pull request fixing the above-mentioned issues and configuring AppVeyor, however, a project owner will need to create the AppVeyor account to grant GitHub access.

It would work by an owner manually creating a GitHub release or tag, AppVeyor would pull and build the code, run NSIS and then upload the resulting installer to the GitHub release. To build locally you would need to install NSIS and the 2 plugins as needed currently, then run the package command, providing the version numbers from git.

**Comment by MarkSummerville** [comment]
Hi @JoeBiellik 

Massive fan of you fixing up the setup of the .Net Framework and setup of AppVeyor as well as any guidance on NuGet/GitHub/Build and release process. 

As a bit of background, @sstteevvee built the app around 2007-2011 and it has existed in a fairly steady state since then.  @mhurwood  and I kept our own branches going, tinkering occasionally. Recently with some changes to TVDB.com's API a few of us have dusted it off, moved to GitHub, developed new websites and supporting structures and have started to actively develop it.  Speaking for myself then I am not a professional developer, so can tinker with the code, but the core setup is a bit beyond my skill level.

In short, I'll get onto AppVeyor now

**Comment by MarkSummerville** [comment]
Hi,

I think I have it up and working with your new yaml file.

There is one error reported:

```
> Update-AppveyorBuild -Version "$($env:APPVEYOR_REPO_TAG_NAME.TrimStart("v"))"
> You cannot call a method on a null-valued expression.
> At line:1 char:34
> + ... ppveyorBuild -Version "$($env:APPVEYOR_REPO_TAG_NAME.TrimStart("v"))"
>     + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
>     + FullyQualifiedErrorId : InvokeMethodOnNull
```

Further the version number doesn't seem to be quite right.

```
Found artifact 'Installer\TVRename-1.0.2.exe' matching 'Installer\TVRename-1.0.2.exe' path
Uploading artifacts...
[1/1] Installer\TVRename-1.0.2.exe (839,311 bytes)...100%
"GitHub" deployment has been skipped as environment variable has not matched ("appveyor_repo_tag" is "false", should be "true")
Updating build cache...
Cache 'packages' - Updated
Build success

```
But it seems to work! yay!

Can you access the AppVeyor for the repository now or do you need me to make you an admin?

Cheers

**Comment by JoeBiellik** [comment]
That's great to hear. I've been using TVRename for many years myself and made my own modifications from a very old copy of the source which I have been using until now.

I see you found my fork, but I don't see which repo you created an AppVeyor account, I don't see it for [this repo](https://ci.appveyor.com/project/TV-Rename/tvrename). 

Moving to this scheme of deployment does mean that the project needs to fully adhere to [semantic versioning](https://semver.org/), with each tag name being at least ``X.Y.Z`` format.

``$APPVEYOR_REPO_TAG_NAME`` in AppVeyor is filled from the git tag name you create the release for. Using the ``appveyor.yml`` from my repo, no changes should need to be made in the AppVeyor settings.

**Comment by JoeBiellik** [comment]
To further clarify, AppVeyor has its own version number it stores and increments (which is why you got 1.0.2), but we are ignoring this, and setting it to the tag name from git.

**Comment by MarkSummerville** [comment]
Happy to move to x.y.z format from 2.3 onwards.

Seems like AppVeyor is under my account https://ci.appveyor.com/project/MarkSummerville/tvrename/settings although it is building the main TV-Rename/master branch. Not sure how to create it against the TV-Rename account though. I'll carry on having a fiddle

**Comment by JoeBiellik** [comment]
Other than checking errors, there should be no reason to check the AppVeyor site, so it shouldn't matter.

All that should now be required to use it, is to make a GitHub release. You create one with a tag of say ``2.3.0`` and publish it, AppVeyor will then use that as the version number to build and upload ``TVRename-2.3.0.exe`` to the release.

Unfortunately, I don't know an easy way to test this without creating a real release on this repo. Perhaps try on your own fork as I did.

Now a CI system is in place it would be trivial to setup automated tests or even nightly or per-commit builds.

The changes you merged from my fork were as minimal as I could make them. If I may, I'll take a look at rewriting some of the NSIS installer and version number related code, keeping functionality but cleaning up.

**Comment by MarkSummerville** [comment]
OK - lets give it a go with the latest beta version. We can create a real release for this version. Stay tuned! Feel free to clean up any of the code, (and fix bugs and add new features!)

**Comment by JoeBiellik** [comment]
A couple of notes: AppVeyor doesn't start until the release is published (drafts won't work) and you need to replace the secret string in ``appveyor.yml`` with one which has access to this repo (see [docs](https://www.appveyor.com/docs/deployment/github/#provider-settings), need to create a GH token, then use AppVeyor tool to secure it).

**Comment by MarkSummerville** [comment]
Thanks - I had just found that and was working on it.  Also worth checking in with @mhurwood who was in the process of updating the installer too

**Comment by MarkSummerville** [comment]
Great work we seem to have a build/release: https://github.com/TV-Rename/tvrename/releases/tag/2.3.6 (no not mention RC1 that very quickly was removed!)

**Comment by JoeBiellik** [comment]
Very nice! The only issue is the installer name and software will report version 2.3.6 as that's the tag you used, it also makes it quite hard to jump to the correct versions code in git.

For ``2.3 RC 2``, semantic versioning states ``2.3.0-rc2`` is the correct tag name, however, that is not a valid Windows version number (4 sections, digits only) which confuses the issue. If you used that tag then AppVeyor will error when it tries to build the project.

I'll have a look at updating the AppVeyor script, it should be possible to trim everything after a dash for the actual file version number, but keep the ``-rc2`` for the filename and display inside the app. Perhaps for now it would be better if you removed the tag or recreated it as ``2.3.0``. Not sure how you would tag RC3 or the final 2.3 release though?

**Comment by MarkSummerville** [comment]
OK - I'll be lead by you and @mhurwood in this regard. In time it would be good to have a nightly/commit level build too so people can get the very latest as soon as we commit it.  How would that work? AppVeyor track the individual build 2.3.0.47 => 2.3.0.48 (tagged as 2.3.0-beta.1) => 2.3.0.49 => 2.3.0.50 => 2.3.0.51 (tagged as 2.3.0-rc.1_ etc?

**Comment by MarkSummerville** [comment]
At the moment I am chuffed to bits that we've got it this far and it seems to be working!

**Comment by MarkSummerville** [comment]
At the moment merges into other branches eg `Andy-ABTec-gh-pages` branch are initiating builds too. Is there a way to restrict builds to just the master branch?

**Comment by Andy-ABTec** [comment]
Happy to make changes at my end if I need to - I'm still "finding my way round" here...

**Comment by JoeBiellik** [comment]
@MarkSummerville I presume you mean Andy's merges are kicking off a build in AppVeyor that fails when it tries to upload to GitHub? AppVeyor already has ``skip_non_tags: true`` which should skip any builds that aren't started from a tagged commit, so I'm not sure whats happening. Could you give me more information?

**Comment by JoeBiellik** [comment]
[Now I see what you mean](https://ci.appveyor.com/project/MarkSummerville/tvrename/build/1.0.14), I'll work on a fix for this.

**Comment by MarkSummerville** [comment]
Cool - good work renaming the branch BTW. Just for my info was this command line wizardry? Do I need to  embrace the console? :-)

**Comment by JoeBiellik** [comment]
For the more advanced git stuff, you really do need to do it CLI, but to rename a branch you can use [GitHub Desktop](https://desktop.github.com/), its a middle ground between the site and the console. I checked out out the old branch, created a new one from it, pushed that to GH and removed the old one but it's easy to [do from the CLI](https://gist.github.com/lttlrck/9628955).

**Comment by JoeBiellik** [comment]
I can't easily test it, but [I think I've got it to skip branches other than master](https://github.com/TV-Rename/tvrename/blob/master/appveyor.yml#L8-L10).

**Comment by MarkSummerville** [comment]
New automated build used on v2.3, so closing
